File: skills/java-superpowers-contract/scripts/sql-explain-analyzer.js
#!/usr/bin/env node
/**
 * SQL Explain Analyzer (Node.js 版)
 * 分析 MySQL 查询执行计划，识别性能瓶颈。
 * 用法: node sql-explain-analyzer.js --db mydb "SELECT * FROM user"
 */
const fs = require("fs");
const { execSync } = require("child_process");
const { MySQLQuery } = require("./database-query.js");

function analyzeExplain(explainData) {
  const issues = [];
  if (explainData && typeof explainData === "object") {
    if (explainData.type === "ALL") issues.push("全表扫描 (type=ALL): 需要添加索引");
    const extra = (explainData.Extra || "").toString();
    if (extra.includes("Using temporary")) issues.push("使用了临时表: 考虑优化GROUP/ORDER BY");
    if (extra.includes("Using filesort")) issues.push("使用了文件排序: 需要添加排序索引");
    if (!explainData.key && explainData.possible_keys) issues.push(`存在可能索引但未使用: ${explainData.possible_keys}`);
  }
  return issues;
}

function parseArgs() {
  const args = process.argv.slice(2);
  const opts = { host: "localhost", port: "3306", user: "root", ssl: "false" };
  let sql = "";
  for (let i = 0; i < args.length; i++) {
    switch (args[i]) {
      case "--host": opts.host = args[++i]; break;
      case "--port": opts.port = args[++i]; break;
      case "--db": opts.db = args[++i]; break;
      case "--user": opts.user = args[++i]; break;
      case "--password": opts.password = args[++i]; break;
      case "--ssl": opts.ssl = args[++i]; break;
      case "--input": opts.input = args[++i]; break;
      default: sql = (sql ? sql + " " : "") + args[i];
    }
  }
  opts.sql = sql;
  return opts;
}

function run(opts) {
  if (opts.input) {
    const content = fs.readFileSync(opts.input, "utf-8");
    content.split("\n").forEach(line => {
      if (/(SELECT|INSERT|UPDATE|DELETE)\s/i.test(line)) {
        console.log(`分析: ${line.trim().slice(0, 80)}...`);
      }
    });
    return;
  }
  if (opts.sql) {
    const cmd = `python database_query.py --host ${opts.host} --port ${opts.port} --db ${opts.db} --user ${opts.user}${opts.password ? " --password " + opts.password : ""} --ssl ${opts.ssl} --explain "${opts.sql.replace(/"/g, '\\"')}"`;
    try {
      const out = execSync(cmd, { timeout: 30000, shell: true }).toString();
      console.log(JSON.stringify(JSON.parse(out), null, 2));
    } catch (e) { console.error("执行失败:", e.message); }
  } else {
    console.log("用法: node sql-explain-analyzer.js --db mydb [--host localhost] [--port 3306] \"SELECT * FROM table\"");
  }
}

run(parseArgs());

File: skills/java-superpowers-contract/scripts/sql_explain_analyzer.py

#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
SQL Explain Analyzer (Python 版)
分析 MySQL 查询执行计划，识别全表扫描、索引使用、临时表等性能瓶颈。
用法：
  python sql_explain_analyzer.py --host localhost --db mydb "SELECT * FROM user WHERE id = 1"
  python sql_explain_analyzer.py --input slow_query.log
"""
import sys
from pathlib import Path
sys.path.insert(0, str(Path(__file__).resolve().parent))

import json, sys, re, os, argparse
from pathlib import Path
try:
    from database_query import MySQLQuery
except ImportError:
    print("错误：缺少 pymysql 依赖。请执行: pip install -r requirements.txt")
    sys.exit(1)

def analyze_explain_json(explain_data):
    issues = []
    if isinstance(explain_data, dict):
        for key in ["key", "possible_keys", "Extra", "type"]:
            val = explain_data.get(key, "")
            if key == "type" and val == "ALL": issues.append("全表扫描 (type=ALL): 需要添加索引")
            if key == "Extra":
                if "Using temporary" in str(val): issues.append("使用了临时表 (Using temporary): 考虑优化GROUP/ORDER BY")
                if "Using filesort" in str(val): issues.append("使用了文件排序 (Using filesort): 需要添加排序索引")
                if "Using index condition" not in str(val) and explain_data.get("type") == "ref": pass
        if not explain_data.get("key") and explain_data.get("possible_keys"):
            issues.append(f"存在可能索引但未使用: {explain_data['possible_keys']}")
        for child in explain_data.get("children", []): issues.extend(analyze_explain_json(child))
        for child in explain_data.get("attached_subqueries", []): issues.extend(analyze_explain_json(child))
    return issues

def run_explain(host, port, db, user, password, sql, ssl_mode):
    q = MySQLQuery(host, port, db, user, password, ssl_mode)
    q.connect()
    result = q.explain(sql)
    q.close()
    return result

def main():
    parser = argparse.ArgumentParser(description="SQL Explain 查询计划分析器")
    parser.add_argument("--host", default="localhost"); parser.add_argument("--port", type=int, default=3306)
    parser.add_argument("--db", required=True, help="数据库名"); parser.add_argument("--user", default="root")
    parser.add_argument("--password", default=os.environ.get("DB_PASSWORD", ""))
    parser.add_argument("--ssl", default="false", choices=["false","true","verify-ca"])
    parser.add_argument("--input", help="慢查询日志文件")
    parser.add_argument("sql", nargs="?", help="待分析的SQL语句")
    args = parser.parse_args()
    if args.input:
        with open(args.input) as f:
            for line in f:
                if re.search(r'(SELECT|INSERT|UPDATE|DELETE)\s', line, re.I):
                    print(f"分析: {line.strip()[:80]}...")
                    out = run_explain(args.host, args.port, args.db, args.user, args.password, line.strip(), args.ssl)
                    data = json.loads(out)
                    issues = analyze_explain_json(data.get("explain", {}))
                    if issues: print("  问题: " + "; ".join(issues))
                    else: print("  状态: 正常")
    elif args.sql:
        out = run_explain(args.host, args.port, args.db, args.user, args.password, args.sql, args.ssl)
        print(json.dumps(out, ensure_ascii=False, indent=2))
    else:
        parser.print_help()
if __name__ == "__main__": main()

